File: java/com/lc/lc199.java
package lc;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class lc199 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Deque<TreeNode> q = new ArrayDeque<>();
        if (root == null) {
            return list;
        }
        q.offer(root);
        while (!q.isEmpty()) {
            int cnt = q.size(), temp = 0;
            for (int i = 0; i < cnt && !q.isEmpty(); i++) {
                TreeNode cur = q.poll();
                temp = cur.val;
                if (cur.left != null) {
                    q.offer(cur.left);
                }
                if (cur.right != null) {
                    q.offer(cur.right);
                }
            }
            list.add(temp);
        }
        return list;
    }
}

File: go/lc/lc199.go
package main


func rightSideView(root *TreeNode) []int {
	res, cur := make([]int, 0), make([]*TreeNode, 0)
	if root == nil {
		return res
	}
	cur = append(cur, root)
	for len(cur) > 0 {
		next := make([]*TreeNode, 0)
		temp := make([]int, 0)
		for _, now := range cur {
			temp = append(temp, now.Val)
			if now.Left != nil {
				next = append(next, now.Left)
			}
			if now.Right != nil {
				next = append(next, now.Right)
			}
		}
		cur = next
		res = append(res, temp[len(temp) - 1])
	}
	return res
}
